src/shared/pages/welcomePage/script.js

```diff
@@ -97,7 +100,11 @@ export async function checkFirefoxHotkeys() {
       await getExternalBrowser(),
     ]);
   } else {
-    replaceMessage(p, "welcomePageNoShortcutsFirefox", "addons://shortcuts/shortcuts");
+    replaceMessage(
```

Comment:
In general, try to avoid unrelated refactors / code formatting changes in patches in which you are expecting review. To perform a code review,  the reviewer has to read the changed code and evaluate whether the change makes sense. A large patch with unrelated changes increases the complexity of the review.

Automatic code formatting changes require less review efforts than code changes that affect the behavior of the program. By separating unrelated changes, the review task is broken down in smaller parts that can enable faster reviews. Note that smaller is not always better: if a patch is split in a non-intuitive way without advance overview to guide the review, then the splitting does more harm than good as it increases the complexity without benefit.

The creation of appropriately sized patches is a skill that requires practice, and does not only benefit reviewers, but also people (including your future self) who need to trace the history of a specific part of the code (e.g. through `git blame`). After all, it is easier to try to understand a self-contained patch that addresses one concern in 50 lines of code than a 500-line patch that mixes many unrelated things. Additionally, if a patch is clearly a mechanical reformatting of code, it is easier to continue tracing the history of the code without wasting too much efforts.

Before the very first review, I asked for one large PR with all changes, which you did (#18). Although definitely not a small patch, that request does not conflict with the above advice: I did not have any prior context on the code in this project, so I eventually had to read through all code. While my life as a reviewer would have been simpler if the changes consisted of smaller patches, I also recognized the reality that it would take lots of time to do that restructuring on your end, and that it would not be unlikely for the result to be more difficult to review than what was given. Fortunately, the overall size was manageable for me, and the readability of the patch was good due to the organization of the code by functionality in reasonably sized files.

Reply:
Thanks for the advice! And yeah, I felt the patch size creeping up on me as I tried to make the webpack build an actual working extension. In terms of the auto formatter, I notice AMO forces proper formatting in the CI checks, which I now realize could have prevented an unintended auto linting of past work not related to this patch.

In terms of the rest of the patch (just so I know for future application), should I have done separate patches as, for example:

1. Replace the webextensions polyfill with the `globalThis.browser ??= chrome;`. Then...
2. Implement the environment variables during the build process. Then...
3. After the review of those, finally focus on replacing the build.js file using webpack so the patch only focuses on copying files and aliasing imports?


Would the sizes of these be appropriate? I would still feel that the 3rd could be considered quite hefty.

Reply:
I would have separated the PR in two parts: one that replaces webextension-polyfill with the simpler one, and one that fixes the build process. I would split them because the changes relate to two independent areas, and it is likely that review+fixups on one area may complete much sooner than the other one (e.g. one approved + ready to land vs the other still awaiting resolution of review issues). In this case, there is a partial overlap in shared code (package.json).

There are several workflows that could apply here:

- Current approach: Create one larger PR that combines unrelated functionality. Pro: easier to develop for yourself. Con: increased review complexity.
  - This is the approach I have suggested for trivial fixups, to minimize review overhead.

- Create multiple branches where changes are stacked on each other (main -> featureA -> featureB). Pro: easier to create at first. Con: more development effort, and requires repeated rebase of the featureB branch on the featureA branch whenever you make changes to featureA (along with resolving merge conflicts if any).
  - And on Github, a pull request for featureB to main would include featureA in the big "Files changed" tab, which means that as a reviewer it's difficult to see which changes are specific to the current PR. This concern can be addressed by setting branchA as the base branch of branchB, and then change it to main (after branchA got merged in main): https://stackoverflow.com/questions/24159036/how-to-modify-a-pull-request-on-github-to-change-target-branch-to-merge-into
  - This workflow is best suited for changes that are largely independent. If there are too many merge conflicts, rebasing branchB all the time would have too much overhead.

- Common workflow, especially in teams: Independent branches: Create separate branches, branched off the common main branch, where you develop the features independently. When one of the two is merged, rebase the other to the (updated) main branch and resolve merge conflicts. Pro: scales better (especially in a team), no need to fix up different branches. Con: there is no single branch with all changes, which reduces the ease of previewing everything.

- Avoiding conflicts at all: Split the work in branchA+branchB in two parts: first the refactor that introduces the new polyfill in branchA, then work in branchB, and after any of them finishes, drop the old polyfill from the shared piece of code.